### ISSUE TITLE
Custom `this` context passed to component upon load.

### ISSUE THREAD

**Issue by philcockfield** [philcockfield/ui-harness#21]
Initial use case: Redux

>> redux just relies on `this.context.store`, which exposes `getState: function, dispatch: function` - @frederickfogerty

----

React context ref: https://facebook.github.io/react/docs/context.html

**Comment by philcockfield** [comment]
For the design of this, I'm thinking we extend the DSL with a `context` so we can declare things like this:


Simple object declartion.

    describe('My Thing', function() {
      context({ ... }); 
    });

Generate the context on demand:

    describe('My Thing', function() {
      context(() => ... ); 
    });

**Comment by philcockfield** [comment]
I'm thinking that we also pass the context into the `this.load()` method, for imperative calls, but when we are specifying the components in the `describe` block function (as per https://github.com/philcockfield/ui-harness/issues/23):

```js
describe(<MyComponent/>, function() {
  context({
    store: {
      getState() {},
      dispatch() {},
    }
  })
});

```

Then we will want a functionally declarative way of expressing the context.

Thoughts @frederickfogerty ?

**Comment by frederickfogerty** [comment]
I was just initially thinking setting context just like the props are set at the moment, with a `this.context` call. I envisage changing context just as often as changing props.

Another option is to change the api completely and instead of calling `this.props` and `this.context` we expect the function signature to be `it :: (props: Object, context: Object) -> ({ props: Object, context: Object })`, which would make the whole api more declarative

To illustrate this better, with the old api (taken from ui-harness-sample):
```js
section("color", () => {
    it("red", () => (this.props({ color: "red" })), this.context({ getState: () => ({ text: "red text" }) }));
    it("green", () => (this.props({ color: "green" })), this.context({ getState: () => ({ text: "green text" }) }));
  });
```

New api:
```js
section("color", () => {
    it("red", () => ({ props: { color: "red" }, context: { getState: () => ({ text: "red text" }) }}));
    it("green", () => ({ props: { color: "green" }, context: { getState: () => ({ text: "green text" }) }}));
  });
```

**Comment by philcockfield** [comment]
Cool - thats even easier.
Excellent :neckbeard:

**Comment by philcockfield** [comment]
Actually - I'm going to do this in two steps.  First extend the current API (which is chain-able BTW), and then look at exposing the more declarative approach that you're come up with which is (1) very appealing, but (2) not quick as simple as I first thought :smile: 

First step (current API style):

```js
section("color", () => {
  it("red", () => (this.props({ color: "red" }).context({ getState: () => ({ text: "red text" }) }));
  it("green", () => {
    this
      .props({ color: "green" })
      .context({ getState: () => ({ text: "green text" });
  });
});
```


New API:
```js
section("color", () => {
    it("red", () => ({ props: { color: "red" }, context: { getState: () => ({ text: "red text" }) }}));
    it("green", () => ({ props: { color: "green" }, context: { getState: () => ({ text: "green text" }) }}));
});
```

**Comment by philcockfield** [comment]
Dammit - turns out dynamically adding a `context` to propogate to children, without statically declaring it with a `childContextTypes` is a sticky problem.

Hmmm, thinking about this.  Might have to make some Redux specific extensions to pass down within the context.

**Comment by frederickfogerty** [comment]
@philcockfield we can just add a constraint that the context has to be statically defined when loading the component, then it's up to the user to define that

**Comment by frederickfogerty** [comment]
Closing as implemented in #45
